### PR TITLE
Hammerhead should not focus an element in Firefox, if it is invisible (close #442)

### LINKED ISSUE #442: Hammerhead should not focus an element in Firefox, if it is invisible.
Firefox does not focus elements if they or their parents have `display: none` in CSS. Also, if you call `element.focus()` and the element is inside an invisible iframe, the iframe will not be active in top document, and the element will not be active in iframe's document. 

`index.html`:
```html
<html>
<head>
    <title>IFrame in FF</title>
</head>
<body>
    <label><input type="checkbox" checked/>Visible</label>
    <button>Focus input in iframe</button>
    <div>
        <iframe src="/inside-iframe.html"></iframe>
    </div>

    <script type="text/javascript">
        window.addEventListener('focus', function (e) {
            console.log('window.top: ', e.target);
        }, true);
        
        document.querySelector('input').addEventListener('change', function () {
            document.querySelector('div').style.display = this.checked ? 'block' : 'none';
        });
        
        document.querySelector('button').addEventListener('click', function () {
            var frame = document.querySelector('iframe');
            var body  = frame.contentDocument.body;
            
            console.log(document.activeElement);
            console.log(frame.contentDocument.activeElement);
            
            body.querySelector('input').focus();
            
            console.log(document.activeElement);
            console.log(frame.contentDocument.activeElement);
        });
    </script>
</body>
```

`inside-iframe.html`:
```html
<html>
<head>
    <title>Iframe</title>
    <script type="text/javascript">
        window.addEventListener('focus', function (e) {
            console.log('frame.window: ', e.target);
        }, true);
    </script>
</head>
<body>
    <input/>
</body>
</html>
```

### PR DESCRIPTION
**Pull request by LavrovArtem**
\cc @churkin @AndreyBelym 
@VasilyStrelyaev please check test names

### CODE REVIEW

File: test/client/fixtures/sandbox/event/focus-blur-change-test.js [hunk 1]
```diff
@@ -936,3 +943,102 @@ asyncTest('scrolling elements with "overflow=hidden" should be restored after fo
...
+    };
+
+    checkFocus({ display: 'none' })
+        .then(function () {
+            return checkFocus({ visibility: 'hidden' });
+        })
+        .then(start);
+});
+
+asyncTest('focus() must not raise event if element in invisible iframe (GH-442)', function () {
```
Comment (VasilyStrelyaev) [inline test/client/fixtures/sandbox/event/focus-blur-change-test.js]:
```
asyncTest('focus() must not raise the event if the element is in an invisible iframe (GH-442)', function () {
```

File: test/client/fixtures/sandbox/event/focus-blur-change-test.js [hunk 2]
```diff
@@ -936,3 +943,102 @@ asyncTest('scrolling elements with "overflow=hidden" should be restored after fo
         start();
     }, false, true);
 });
+
+asyncTest('focus() must not raise event if element is invisible (GH-442)', function () {
```
Comment (VasilyStrelyaev) [inline test/client/fixtures/sandbox/event/focus-blur-change-test.js]:
```
asyncTest('focus() must not raise the event if the element is invisible (GH-442)', function () {
```

**Comment by VasilyStrelyaev** [comment]
\r-

**Comment by LavrovArtem** [comment]
FPR

**Comment by VasilyStrelyaev** [comment]
lgtm

File: src/client/sandbox/event/focus-blur.js [hunk 3]
```diff
@@ -198,6 +197,10 @@ export default class FocusBlurSandbox extends SandboxBase {
         if (this.shouldDisableOuterFocusHandlers && !domUtils.isShadowUIElement(el))
             return null;
 
+        if (isNativeFocus &&
```
Comment (churkin) [inline src/client/sandbox/event/focus-blur.js]: need comment

**Comment by AndreyBelym** [comment]
lgtm

File: src/client/sandbox/event/focus-blur.js [hunk 4]
```diff
@@ -198,6 +197,10 @@ export default class FocusBlurSandbox extends SandboxBase {
...
+            (!browserUtils.isWebKit && styleUtils.isElementInInvisibleIframe(el) || !styleUtils.isElementVisible(el)))
```
Comment (inikulin) [inline src/client/sandbox/event/focus-blur.js]: Refactor long conditions to the vars

**Comment by inikulin** [comment]
\r-

**Comment by churkin** [comment]
\r-

**Comment by LavrovArtem** [comment]
FPR

**Comment by churkin** [comment]
lgtm

**Comment by LavrovArtem** [comment]
@inikulin FPR

**Comment by inikulin** [comment]
lgtm

**Pull request merged**